http-client-core/src/main/java/io/micronaut/http/client/bind/binders/QueryValueClientArgumentRequestBinder.java

```diff
@@ -0,0 +1,400 @@
+     *     <tr> <td> Object </td>           <td> param={name: "Mike", age: 30} </td>               <td> "param[name]=Mike&param[age]=30" </td> </tr>
+     * </table>
+     */
+    public static final String QUERY_VALUE_FORMAT_DEEP_OBJECT = "DEEP_OBJECT";
+
+    private static final String CSV_DELIMITER = ",";
+    private static final String SSV_DELIMITER = " ";
+    private static final String PIPES_DELIMITER = "|";
+
+    private final ConversionService<?> conversionService;
+
+    public QueryValueClientArgumentRequestBinder(ConversionService<?> conversionService) {
+        this.conversionService = conversionService;
+
+        // Add the converters
+        conversionService.addConverter(Iterable.class, ConvertibleMultiValues.class, new IterableConverter());
+        conversionService.addConverter(Map.class, ConvertibleMultiValues.class, new MapConverter());
+        conversionService.addConverter(Object.class, ConvertibleMultiValues.class, new ObjectConverter());
+    }
+
+    @Override
+    @NonNull
+    public Class<QueryValue> getAnnotationType() {
+        return QueryValue.class;
+    }
+
+    /**
+     * If value can be converted to ConvertibleMultiValues, then use it and add it to the uriContext.queryParameters.
+     * Otherwise if the {@link Format} annotation is present, it is converted to {@link String}. If none of these
+     * are satisfied, the{@link io.micronaut.http.uri.UriTemplate} decides what to do with the given value which
+     * is supplied as an Object (it is added to uriContext.pathParameter).
+     *
+     * <br> By default value is converted to ConvertibleMultiValues when the {@link Format} annotation is present and has
+     * one of the defined above formats. Otherwise empty optional is returned.
+     *
+     * <br> The default {@link io.micronaut.http.uri.UriTemplate} will convert the value to String and to parameters.
+     * Optionally, the value can be formatted if the path template states so.
+     */
+    @Override
+    public void bind(
+            @NonNull ArgumentConversionContext<Object> context,
+            @NonNull ClientRequestUriContext uriContext,
+            @NonNull Object value,
+            @NonNull MutableHttpRequest<?> request
+    ) {
+        ArgumentConversionContext<ConvertibleMultiValues> conversionContext = ConversionContext.of(
+                Argument.of(ConvertibleMultiValues.class, context.getArgument().getName(), context.getAnnotationMetadata()));
+        Optional<ConvertibleMultiValues> convertedValue = conversionService.convert(value, conversionContext);
+
+        if (convertedValue.isPresent()) {
+            // TODO How should I check if it is truely a List of Strings?
+            ConvertibleMultiValues<String> multiValues;
+            try {
+                // noinspection unchecked
+                multiValues = convertedValue.get();
```

Comment:
.get() should never throw an exception if `isPresent()` is true

Reply:
The try-catch was in case the type was some ConvertibleMultiValues<T> and not ConvertibleMultiValues<String>, but since java won't throw an exception in this case, I will probably have to check the values inside.